catalog/catalog.go

```diff
@@ -47,19 +52,136 @@ func WithAwsConfig(cfg aws.Config) Option {
 	}
 }
 
+func WithCredential(cred string) Option {
```

Comment:
It feels like these options are being overloaded with probably only minimal overlap, do you think we should move these catalog implementations into their own packages?

I got a feeling this would happen with "common options", interested in your thoughts.

Reply:
Yea I was thinking the same. Though my question is how many of these options will overlap with the other catalog implementations. Several of them probably could overlap. (This is probably why all the options in pyiceberg are all passed via property mappings)

Reply:
@zeroshade yeah I think it is important to remember that a little duplication is OK, as long as the user of the API gets a neat typed interface.

Sure internally there maybe some overlap, but that typically erodes over time as corner cases evolve for each implementation. 

As a user of the API I am using Glue + S3, am i likely to switch to hive + REST? 

Personally I prefer a typed interface in Go APIs, over a bag of properties.

Reply:
@wolfeidau take a look at the way I've set up some interesting little generic stuff in the most recent commit making the `Options` typed so that we can specify options which are explicit for one catalog or the other without having to do a ton of duplication. Let me know what you think.